File: 2025_06/leetcode2025-06-16.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

class Solution {
public:
    int maximumDifference(vector<int>& nums) {
        int n = nums.size();
        int res = -1;
        int lastMin = INT_MAX;
        for (int i=0; i<n; i++){
            res = max(res, nums[i]-lastMin);
            lastMin = min(lastMin, nums[i]);
        }
        return res==0 ? -1 : res;
    }
};

int main(){
    Solution s;
    vector<int> nums = {7,1,5,4};
    cout << s.maximumDifference(nums) << endl;
}

File: leetcode2025/2025_06/leetcode2025-06-16.py
from typing import List

class Solution:
    def maximumDifference(self, nums: List[int]) -> int:
        # 对于每一个位置i，我们只需要知道nums[:i]的最小值即可求出以i为结尾的最大差值
        # 所以我们可以在循环中始终维护一个nums[:i]的最小值
        myMax = lambda x,y : x if x>y else y
        myMin = lambda x,y : x if x<y else y
        res = -1
        lastMin = float("inf")
        n = len(nums)
        for i in range(n):
            res = myMax(res, nums[i]-lastMin)
            lastMin = myMin(lastMin, nums[i])
        return -1 if res == 0 else res

        # res = -1
        # n = len(nums)
        # for i in range(n):
        #     for j in range(i+1, n):
        #         res = max(res, nums[j]-nums[i])
        # return -1 if res == 0 else res

if __name__ == "__main__":
    s = Solution()
    nums = [7,1,5,4]
    print(s.maximumDifference(nums))
